core/src/main/kotlin/rys/ajaxpetproject/service/impl/UserServiceImpl.kt

```diff
@@ -1,39 +1,90 @@
 import rys.ajaxpetproject.repository.UserRepository
 import rys.ajaxpetproject.service.UserService
 
 @Service
+@Suppress("MagicNumber")
 class UserServiceImpl(
-    val userRepository: UserRepository,
-    val passwordEncoder: PasswordEncoder) : UserService {
+    private val encoder: PasswordEncoder,
+    private val userRepository: UserRepository
+) : UserService {
 
-    override fun createUser(mongoUser: MongoUser): MongoUser {
-        return userRepository.save(mongoUser.copy(password = passwordEncoder.encode(mongoUser.password)))
+    override fun createUser(mongoUser: MongoUser): Mono<MongoUser> {
+        return userRepository.findByName(mongoUser.userName)
+            .handle<MongoUser> { _, sync ->
+                sync.error(UserAlreadyExistsException("A user with the username ${mongoUser.userName} already exists!"))
+            }
+            .switchIfEmpty {
+                (mongoUser.password.length > MIN_PASSWORD_LENGTH)
+                    .let {
+                        val nesUser = mongoUser.copy(password = encoder.encode(mongoUser.password))
+                        userRepository.save(nesUser)
+                    }
+                    .switchIfEmpty {
+                        Mono.error(
+                            IllegalArgumentException("Password must be at least $MIN_PASSWORD_LENGTH characters long!")
+                        )
+                    }
+            }
     }
 
-    override fun getUserById(id: ObjectId): MongoUser = userRepository.findUserById(id) ?: throw UserNotFoundException()
+    override fun findUserById(id: String): Mono<MongoUser> {
+        return userRepository.findById(id)
+    }
+
+    override fun findUserByName(name: String): Mono<MongoUser> {
+        return userRepository.findByName(name)
+    }
+
+    override fun getUserById(id: String): Mono<MongoUser> {
+        return userRepository.findById(id)
+            .switchIfEmpty {
+                Mono.error(UserNotFoundException("User with id $id does not exist!"))
+            }
+    }
 
-    override fun findUserById(id: ObjectId): MongoUser? = userRepository.findUserById(id)
+    override fun getUserByName(name: String): Mono<MongoUser> {
+        return userRepository.findByName(name)
+            .switchIfEmpty {
+                Mono.error(UserNotFoundException("User with name $name does not exist!"))
+            }
+    }
 
-    override fun findAllUsers(): List<MongoUser> = userRepository.findAllBy()
+    override fun findAllUsers(): Flux<MongoUser> {
+        return userRepository.findAll()
+    }
 
-    override fun updateUser(id: ObjectId, updatedMongoUser: MongoUser): MongoUser? =
-        findUserById(id)?.let {
-            userRepository.save(updatedMongoUser.copy(id = id))
-        } ?: throw UserNotFoundException()
+    override fun updateUser(id: String, updatedUser: MongoUser): Mono<MongoUser> {
+        return getUserById(id)
+            .flatMap {
+                userRepository.update(
```

Comment:
why do you need two queries to the DB to update user?

Reply:
Main purpose of first query is to get User's Id. 
1)This approach gives more control on Update operation. (For example : prohibit update for some users)
3)Id creation is MongoDB responsibility so updatedUser normally can't have Id

Reply:
What if we execute query like:
```sql
update users 
  set field1=:value1, field2=:field2 
where userId=:userId
```

Mongo will return update result from which you can find out whether the query has updated something.

----

What's wrong with your approach. Consider the next events:
1. user with an ID exists
2. [Thread-1] you received the user using `getUserById` function
3. [Thread-2] some concurrent thread removed the user
4. [Thread-1] `userRepository.update` is invoked


---- 

Nevermind. Just mentioned a possible issue in your code. Feel free to resolve the thread

Reply:
That's in fact issue. 
I really understood why we **must** to decrease amount of queries. 
I remembered this experience. 
For now I'll leave it as it is and maybe will return later when I have time to fix this